proof/crefine/RISCV64/Finalise_C.thy

```diff
@@ -187,29 +182,55 @@ lemma tcbSchedEnqueue_cslift_precond_discharge:
   apply (clarsimp simp: ctcb_relation_def)
   apply (frule arg_cong[where f=unat], subst(asm) unat_ucast_up_simp, simp)
   apply (frule tcb_queue'_head_end_NULL)
-   apply (erule conjunct1[OF valid_queues_valid_q])
+   apply (fastforce dest!: valid_queues_valid_q)
   apply (frule(1) tcb_queue_relation_qhead_valid')
    apply (simp add: valid_queues_valid_q)
   apply (clarsimp simp: h_t_valid_clift_Some_iff)
   done
 
+lemma refill_unblock_check_ccorres:
+  "ccorres dc xfdc (invs' and sc_at' scPtr) \<lbrace>\<acute>sc = Ptr scPtr\<rbrace> []
+    (refillUnblockCheck scPtr) (Call refill_unblock_check_'proc)"
+sorry (* FIXME RT: refill_unblock_check_ccorres *)
+
 lemma cancel_all_ccorres_helper:
   "ccorres dc xfdc
-      (\<lambda>s. valid_objs' s \<and> valid_queues s
-             \<and> (\<forall>t\<in>set ts. tcb_at' t s \<and> t \<noteq> 0)
-             \<and> sch_act_wf (ksSchedulerAction s) s)
-      {s'. \<exists>p. ep_queue_relation (cslift s') ts
-                p (thread_' s')} hs
-      (mapM_x (\<lambda>t. do
-              y \<leftarrow> setThreadState Restart t;
-              tcbSchedEnqueue t
-            od) ts)
-      (WHILE \<acute>thread \<noteq> tcb_Ptr 0 DO
-        (CALL setThreadState(\<acute>thread, scast ThreadState_Restart));;
-        (CALL tcbSchedEnqueue(\<acute>thread));;
-        Guard C_Guard \<lbrace>hrs_htd \<acute>t_hrs \<Turnstile>\<^sub>t \<acute>thread\<rbrace>
-         (\<acute>thread :== h_val (hrs_mem \<acute>t_hrs) (Ptr &(\<acute>thread\<rightarrow>[''tcbEPNext_C'']) :: tcb_C ptr ptr))
-       OD)"
+     (\<lambda>s. valid_objs' s \<and> valid_queues s
+          \<and> (\<forall>t\<in>set ts. tcb_at' t s \<and> t \<noteq> 0)
+          \<and> sch_act_wf (ksSchedulerAction s) s)
+     {s'. \<exists>p. ep_queue_relation (cslift s') ts p (thread_' s')} hs
+     (mapM_x cancelAllIPC_loop_body ts)
+     (WHILE \<acute>thread \<noteq> NULL DO
+      Guard C_Guard \<lbrace>hrs_htd \<acute>t_hrs \<Turnstile>\<^sub>t \<acute>thread\<rbrace>
+            (\<acute>ret__unsigned_longlong :== CALL thread_state_get_replyObject(h_val (hrs_mem \<acute>t_hrs)
```

Comment:
I see what you meant with large ugly chunk of SIMPL.. I think it's Ok to leave this here for now, but there is a good chance that we might need to break this up further somehow when we get to the proof.